### ISSUE TITLE
Is redeclaration necessary to change a parameter value in a class?

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#2149]
**Modified by sjoelund&#x2e;se on 12 Oct 2017 13:49 UTC**
The following is a simplification of some code found in a Modelica library:

```mo
model A
    replaceable model S
	parameter Integer n;
	Real x;
    equation
        x = n;
    end S;
    S s;
end A;

model B
    extends A(S(n = 2));
end B;
```

Is the **extends** clause in model B legal?

The specification states that redeclarations are used to replace the declaration of a class or component with another, but does not state that a redeclaration _must_ be used for this.

Likewise, when talking about the `{X(Y(z=2))}` syntax, it doesn't state whether X and Y can be classes, or only components.

----

**Reported by stefanv on 19 Jan 2017 19:04 UTC**
The following is a simplification of some code found in a Modelica library:

```modelica
model A
    replaceable model S
	parameter Integer n;
	Real x;
    equation
        x = n;
    end S;
    S s;
end A;

model B
    extends A(S(n = 2));
end B;
```

Is the **extends** clause in model B legal?

The specification states that redeclarations are used to replace the declaration of a class or component with another, but does not state that a redeclaration _must_ be used for this.

Likewise, when talking about the `{X(Y(z=2))}` syntax, it doesn't state whether X and Y can be classes, or only components.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/2149

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 12 Oct 2017 08:31 UTC**
The macro-processor for Modelica seems broken, the example is:
```
 model A
     replaceable model S
         parameter Integer n;
         Real x;
     equation
         x = n;
     end S;
     S s;
 end A;

 model B
     extends A(S(n = 2));
 end B;
```
This syntax is used a lot in Modelica.Media (e.g. modifying the Temperature-type); so we should ensure that it is legal in those cases.

However, there is a subtle issue: the Temperature is a short class, and it could that we only need this in those cases (and for modifiers they already work slightly differently).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 12 Oct 2017 10:02 UTC**
To discuss at meeting. Some should be possible to trivially close.

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 12 Oct 2017 10:24 UTC**
I do not understand the issue. For me modification and redeclaration are two separate concepts, with modification about changing values and redeclaration about changing types (which of course in turn can change values or introduce new variables and equations).

**(1)** Said in other words, why should it not be possible to extend a model and modify one of its replaceable components without redeclaring it, as given in your example:
```
model A
  replaceable model S
    parameter Integer n;
    Real x;
  equation
    x = n;
  end S;
  S s;
end A;

model B
  extends A(S(n = 2));
end B;
```

**(2)** Or the other way around, why should it not be possible to extend a model and redeclare one of its replaceable components without modifying:
```
package P
  model A
   replaceable model S
       parameter Integer n = 40;
       Real x;
   equation 
       x = n;
   end S;
   S s;
  end A;

  model B
    extends A;
    redeclare model S
       parameter Integer n = 30;
       Real x;
    equation 
       x = n + 3;
    end S;
  end B;
end P;
```

**(3)** The really ugly one is doing both, where I am not sure different Modelica compilers agree upon or its even clear in the standard what actually happens:
```
package P
  model A
   replaceable model S
       parameter Integer n = 40;
       Real x;
   equation 
       x = n;
   end S;
   S s;
  end A;

  model B
    extends A(S(n = 20));
    redeclare model S
       parameter Integer n = 30;
       Real x;
    equation 
       x = n + 3;
    end S;
  end B;
end P;
```
In Dymola simulation of above model yields `P.B.x = 33`, even when we switch the `extends` and `redeclare` order.

**(4)** Using short-class-definitions, case **(3)** can be made clear:
```
package P
  model A
   replaceable model S
       parameter Integer n = 40;
       Real x;
   equation 
       x = n;
   end S;
   S s;
  end A;

  model SS
    parameter Integer n = 20;
      Real x;
  equation 
      x = n + 3;
  end SS;

  model B
    extends A(redeclare model S = SS(n = 10));
  end B;
end P;
```
Simulation yields `P.B.x = 13`.

Maybe case **(3)** needs clarification; should we denote it as bad style, give a warning or even be it an error?

**Comment by modelica-trac-importer** [comment]
**Modified by sjoelund&#x2e;se on 12 Oct 2017 13:49 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 12 Oct 2017 15:55 UTC**
Related to **(3)**:
The (new) text in the specification 3.4 chapter 5.6.1.4 reads:

 A partially instantiated class or component is an element that is ready to be instantiated; a partially instantiated element (i.e. class or component) is comprised of a reference to the original element (from the class tree) and the modifiers for that element (including a possible redeclaration).

 The possible redeclaration of the element itself takes effect.

 The class of a partially instantiated component is found in the instance tree (**using the redeclaration** if any), **modifiers merged to that class** forming a new partially instantiated class that is instantiated as below.

For me, it means redeclarations are performed first and then modifications are applied.
I.e. when P.B.s ist instantiated its class is found in the redeclaration `redeclare model S`and the applicable modifiers `(S(n=20))`are merged and applied to that class.

Therefore the result should be 23.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 13 Oct 2017 10:08 UTC**
Replying to [comment:5 Gerd Kurzbach]:
> Related to **(3)**:
> [...]
> For me, it means redeclarations are performed first and then modifications are applied.
> I.e. when P.B.s ist instantiated its class is found in the redeclaration `redeclare model S`and the applicable modifiers `(S(n=20))`are merged and applied to that class.
> 
> Therefore the result should be 23.

I disagree. There are two instantiations involved there. First the inherited class `A(S(n = 20))` is instantiated, it has no redeclarations, and its modifier is applied. The second one is `B` - it has a redeclare but other no modifier (but the redeclare has a modifier). Thus the `n = 20` modifier is applied to the replaceable class, and then the modified replaceable class is redeclared with the version with `n = 30`. This would give the same result as Dymola.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 13 Oct 2017 11:12 UTC**
Replying to [comment:6 Jesper Mattsson]:
> I disagree. There are two instantiations involved there. First the inherited class `A(S(n = 20))` is instantiated, it has no redeclarations, and its modifier is applied. The second one is `B` - it has a redeclare but other no modifier (but the redeclare has a modifier). Thus the `n = 20` modifier is applied to the replaceable class, and then the modified replaceable class is redeclared with the version with `n = 30`. This would give the same result as Dymola.

Perhaps I repeat my argumentation a little bit more in detail:

The instantiation starts with B which leads to the instantiation of A as base class of B where the modifiers of A are applied, but not those of S, which are kept in the modification environment. When it then comes to s, already the redeclared class S is used, and the modifiers of S are applied and S.n gets 20. Then, this class is used to instantiate s. So B.s.n gets 20.

Obviously, there are two different interpretations, which leaves room for clarifactions in the spec., but IMO chapter 5.6.1.4 describes this already clearly. Maybe this example should be added.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 16 Oct 2017 09:14 UTC**
Replying to [comment:7 Gerd Kurzbach]:
> [...], but not those of S, which are kept in the modification environment.

Why not? That seems arbitrary to me. Could you point to the part of the spec text that says that they are not applied before the redeclare?

I think that the order of operation that you describe break the logic of modifiers and redeclares. The modifier is applied to the original declaration, and the redeclare replaces the original declaration. If the text specifies that they are applied in the other order, then I'd say it needs to be refined.

Now, there is another argument for the answer being 23 - you could argue that the modification of S makes that modification a part of the constraining type, and that would mean that it is also applied to the replacing class. The latest version of the text of 7.3.2 seems to preclude that interpretation by accident (as it assumes that the replaceable declaration is a component or short class declaration). However, if S was a short class declaration, then it could be taken to mean either that the modifier is or isn't part of the constraining type. (Depending on how you interpret "modifiers on" as used there.)

Consider:
```mo
package P
  model S0
     parameter Integer n = 40;
     Real x;
   equation 
     x = n;
  end S0;

  model A1
    replaceable model S = S0;
    S s;
  end A1;

  model B1
    extends A1(S(n = 20));
    redeclare model S
       parameter Integer n = 30;
       Real x;
    equation 
       x = n + 3;
    end S;
  end B1;

  model A2
    replaceable model S = S0 constrainedby S0;
    S s;
  end A2;

  model B2
    extends A2(S(n = 20));
    redeclare model S
       parameter Integer n = 30;
       Real x;
    equation 
       x = n + 3;
    end S;
  end B2;

  model B3
    extends A1(S(n = 20));
    redeclare model S = S3(n = 30);
  end B3;

  model S3
     parameter Integer n = 50;
     Real x;
   equation 
     x = n + 3;
  end S3;
end P;
```
Here B1 clearly is affected by the ambiguity in 7.3.2, while B2 clearly is not (since it has a constrainedby clause). 

B3 on the other hand would yield `x = 33` with both your interpretation and mine, right? I see B3 and the B from **(3)** as two different ways to express the same thing, and would find it very subtle and confusing if they gave different results.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 17 Oct 2017 09:46 UTC**
Replying to [comment:8 Jesper Mattsson]:
> Replying to [comment:7 Gerd Kurzbach]:
> > [...], but not those of S, which are kept in the modification environment.
> Why not? That seems arbitrary to me. Could you point to the part of the spec text that says that they are not applied before the redeclare?

At this point there is no S yet; when it is instantiated below A, it will get its modifier.

> 
> I think that the order of operation that you describe break the logic of modifiers and redeclares. The modifier is applied to the original declaration, and the redeclare replaces the original declaration. If the text specifies that they are applied in the other order, then I'd say it needs to be refined.

For me, its exactly the other way around: The redeclaration replaces the original declaration and then modifiers are applied. And I find this quite natural. (All other interpretations would break the logic of modifiers and redeclares ;-)

> 
> Now, there is another argument for the answer being 23 - you could argue that the modification of S makes that modification a part of the constraining type, and that would mean that it is also applied to the replacing class. The latest version of the text of 7.3.2 seems to preclude that interpretation by accident (as it assumes that the replaceable declaration is a component or short class declaration). However, if S was a short class declaration, then it could be taken to mean either that the modifier is or isn't part of the constraining type. (Depending on how you interpret "modifiers on" as used there.)

I interpret this sentence in this way that only modifiers placed on the original declaration are applied automatically on the redeclared element. So it has no influence here.

> 
> Consider:
> ...
> Here B1 clearly is affected by the ambiguity in 7.3.2, while B2 clearly is not (since it has a constrainedby clause). 
> 
> B3 on the other hand would yield `x = 33` with both your interpretation and mine, right? I see B3 and the B from **(3)** as two different ways to express the same thing, and would find it very subtle and confusing if they gave different results.

B1 and B2 yield 23, because the modifiers are applied after redeclaration.
B3 yields 33, but I consider it as a bug in SimulationX and it should be 23 too.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 17 Oct 2017 10:00 UTC**
I agree with Gerd's interpretation above, and that is apparently what I implemented, since MapleSim produces 23 for all three cases.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 17 Oct 2017 10:24 UTC**
I think the whole example points out a fundamental issue with modifications in extends statements: You cannot tell whether the modified class is the one being inherited or the one being redeclared.

In such a situation, I would like to know what the use case for such an example is. Certainly the interpretation of Gerd and Stefan is redundant, since the user could as well replace the definition inside the redeclaration, no? 

On the other hand, **if** a redeclaration is present, the modification is redundant according to Jesper's interpretation as it is overriden by the redeclaration, no?

In conclusion, the only reasonable use case is to modify a replaceable class that is not redeclared in the same scope. Should we simply issue an error in such a case, to avoid the ambiguity?

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 17 Oct 2017 11:42 UTC**
It could work as follows:

From the Spec, modifications which are related to the same instance are handled by merging, where outer modifications override inner. So the loss of a modification is a common case.

Modifications at the class of an instance are always more inner than modifications referring to the component name. From all modifications, these directly at the declaration are the innermost.

In this inner/outer relation graph, when redeclaring a class, its class declaration is replaced by a new class declaration at the innermost place. Therefore, it is common that outer (later) modifications will override modifications inside the class. 

In the example B3, the modification n=30 of S=S3 is moved by redeclaration to the innermost place because it is the declaration of the class itself. Therefore, S(n=20) is outermore (it is not the declaration) and overrides n=30.

With this, there is no ambiguity and no need to forbid it.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 17 Oct 2017 12:31 UTC**
I think you miss the point. Of course you can define this situation in a way that yields an unambiguous solution - at the cost of talking about "moving modifications" and an "inner/outer relation graph" which are AFAIK nowhere specified. And of course, modifications can replace existing modifications, this is again not my point.

My point is that the modification of `S` is ambiguous. Your explanation puts the redeclaration first, followed by the (nested) modification. AFAIK, the specification is unclear about this ordering and one could as well first modify the inherited S and then replace it. But whatever order we decide upon, it simply does not make much sense, since one of the declarations vanishes. The user might be surprised. 

At least, I consider it worth a warning if a base-class modification and redeclarations collide, but I do still not see use-case for it. And before we extend the spec to "make" something "work" that has no use-case, we should simply forbid it. It makes the language easier to implement and maintain IMO.

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 17 Oct 2017 14:27 UTC**
As assumed, tools disagree on **(3)** and the specification needs clarification. I like to add that **(3)** is just a _simple_ instance of a bunch of modifier vs. redeclaration diamond problems (**(3)** is simple because it combines things by a single extends).

The following is not so natural:
 _"The redeclaration replaces the original declaration and then modifiers are applied. And I find this quite natural. (All other interpretations would break the logic of modifiers and redeclares ;-)"_

This interpretation gives modification preference over redeclaration by applying redeclarations first, followed by modifications. It is natural in the sense that first types are settled before values of their instances are modified. On the one hand, it is not natural since something _is preferred by doing something else first_ and because the distinction between type and value is completely unclear not at least due to value definitions being part of models (i.e., types). For me, redeclaration and modification are without any natural order: redeclaration to change type (thus, to modify the type which can include default values and equations) and modification to change value (thus, to modify the value). Why one should have preference over the other is hardly natural, but we could just define an order like first fix types followed by values. The implication is that its pushing quiet a burden on users as they have to apply the standard in their mind to figure out what is going on ... and it is not so simple as shown in the following.

For me, the model of **(3)** is contradicting itself, stating `P.B.A.n` is `30` by redeclaration and at the same time `P.B.A.n` is `20` by modification. But let's assume we do as Gerd suggests and we first settle redeclarations and afterwards modifications, i.e., we first perform type modifications and afterwards value modifications. This is not so simple and for sure not end-user intuitive for two reasons: the diamond problem and conditional declarations in combination with component redeclaration and modification.

**Diamond problem**

**(5)** We can make our example worse when redeclaration and modification of the same component is not done within the same model but by several independent models we extend from, i.e., in case of the diamond problem:
```
package P
  model A
   replaceable model S
       parameter Integer n = 40;
       Real x;
   equation 
       x = n;
   end S;
   S s;
  end A;

  model A1
    extends A;
    redeclare model S
       parameter Integer n = 30;
       Real x;
    equation 
       x = n + 3;
    end S;
  end A1;

  model A2
    extends A(S(n = 20));
  end A2;

  // Should order of extends in the following matter:
  
  model B1
    // No warning in Dymola:
    extends A1;
    extends A2;
  end B1;

  model B2
    // Multiple definitions warning in Dymola:
    extends A2;
    extends A1;
  end B2;
end P;
```
Besides being troublesome to define, the implications for users are complex making it hard to follow what is going on.

Other languages (for example Eiffel) with multiple inheritance often demand explicit disambiguation of contradicting definitions instead of some auto-magic resolution.

**Conditional declarations in combination with component redeclaration and modification**

The simple approach of Gerd and Stefan to just prefer redeclarations (type modifications) followed by modifications (value modifications) is insufficient if component redeclarations and condidtional declarations are involved:
```
package P3
  model A
   replaceable model S
       parameter Integer n = 40;
       Real x;
   equation 
       x = n;
   end S;

   S s1;
   replaceable S s2 if s1.n == 10; // == 30 for Jesper such that component exists
   //S s3 if s2.n == 60; // Rejected by Dymola (to stop insanity :-)
  end A;

  model B
    extends A(S(n = 10), redeclare SS s2(n2 = s1.n));

    redeclare model S
       parameter Integer n = 30;
       Real x;
    equation 
       x = n + 3;
    end S;

    model SS
     extends S(n = n2 + 50);
     parameter Integer n2 = 40;
    end SS;
  end B;
end P3;
```
The problem with this model is that redeclarations and other modifications have to be evaluated intertwined to get to a consistent result with Gerd's interpretation:
 1. Redeclare `S`
 2. Modify its `n` parameter
 3. `s1` can be instantiated with `s1.n = 10`, resulting in `s2` to exist
 4. Redeclare `s2` based on `SS`
 5. Modify `n2` of `s2` (note that `s2.n = 10` due to 2)
This is similar to demand-driven macro processing in Lisp or Scheme; it is far to crazy language-wise for an engineering domain.

Of course we can spice it even more with the diamond problem by splitting component redeclarations and other modifications via multiple inheritance...

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 19 Oct 2017 16:11 UTC**
The case (3) is illegal:

7.3 says: "The redeclare construct as an element requires that the element is inherited, and cannot be combined with a modifier of the same element in the extends-clause."
\--
The original case from Stefan is used in Modelica.Media, is included in 4.5.3 of specification and there is nothing explicitly prohibiting it so it should be legal (with the modifier applied in the expected way).

Agreement.

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 20 Oct 2017 10:06 UTC**
Replying to [comment:15 Hans Olsson]:
> The case (3) is illegal:
> 
> 7.3 says: "The redeclare construct as an element requires that the element is inherited, and cannot be combined with a modifier of the same element in the extends-clause."

It is unclear to me how this restriction handles all cases, particularly multiple inheritance (Diamond problem, simplified case **(5)**):
```
package P
  model A
   replaceable model S
       parameter Integer n = 40;
   end S;
   S s;
  end A;

  model A1
    extends A;
    redeclare model S
       parameter Integer n = 30;
    end S;
  end A1;

  model A2
    extends A(S(n = 20));
  end A2;

  model B1
    // MULTIPLE inheritance modifies the same component INDIRECTLY, i.e.,
    // there are several contradicting paths in the extends graph:
    extends A1;
    extends A2;
  end B1;
end P;
```
Model `A1` does not combine redeclaration and modification, neither does `A2`. The problem happens later, when `B` extends from both. But that is to late for 7.3 to be applicable (there is no `redeclare` or modification in `B`). For the intention of 7.3 to work, we need some transitivity constraint.

**(6)** Please note, that the following should be Ok, as there are no contradicting modifications (neither value nor type), but just a chain thereof:
```
package P
  model A0
   replaceable model S
       parameter Integer n = 40;
   end S;
   S s;
  end A0;

  model A1
    extends A0;
    redeclare replaceable model S
       parameter Integer n = 30;
    end S;
  end A1;

  model A2
    extends A1(s(n = 20));
  end A2;

  model A3
    extends A2;
    redeclare model S
       parameter Integer n = 10;
    end S;
  end A3;

  model B
    extends A3;
  end B;
end P;
```

**Comment by modelica-trac-importer** [comment]
**Modified by cbuerger on 20 Oct 2017 10:11 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 20 Oct 2017 10:59 UTC**
Replying to [comment:9 Gerd Kurzbach]:
> B1 and B2 yield 23, because the modifiers are applied after redeclaration.
> B3 yields 33, but I consider it as a bug in SimulationX and it should be 23 too.

Consider:
```mo
package P
  model S0
     parameter Integer n = 40;
  end S0;

  model A
    replaceable model S = S0;
    S s;
  end A;

  model B
    extends A(S(n = 20));
  end B;

  model C
    extends B(redeclare S = S0(n = 10));
  end C;
end P;
```

As I understand what you are saying, your interpretation means that in C, `s.n = 20`. To me that does not at all seem like what the user would expect. Did I misunderstand your description?

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 10 Jan 2018 08:23 UTC**
